### PR TITLE
Various DX coil fixes

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - This contains various fixes related to the old DX coils taken from the Coil:DX:Cooling work.
 - Use `PLR`, not `RuntimeFraction` to calculate leaving conditions for 2-stage operation (Coil:Cooling:DX:TwoStageWithHumidityControlMode).
 - Fix Multispeed DX massflow weighting between speeds (Coil:Cooling:DX:Multispeed)
 - Fix Multispeed DX outlet conditions at low speed (Coil:Cooling:DX:Multispeed)
 - Allow 2-speed DX to operate as variable speed (Coil:Cooling:DX:TwoSpeed)

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DXCoils.cc [hunk 1]
```diff
@@ -614,8 +614,8 @@ namespace EnergyPlus::DXCoils {
 
                     // Determine combined performance
                     state.dataDXCoils->DXCoil(DXCoilNum).OutletAirEnthalpy =
-                        (1.0 - S12RuntimeFraction) * S1OutletAirEnthalpy + S12RuntimeFraction * S12OutletAirEnthalpy;
-                    state.dataDXCoils->DXCoil(DXCoilNum).OutletAirHumRat = (1.0 - S12RuntimeFraction) * S1OutletAirHumRat + S12RuntimeFraction * S12OutletAirHumRat;
+                        (1.0 - S2PLR) * S1OutletAirEnthalpy + S2PLR * S12OutletAirEnthalpy;
+                    state.dataDXCoils->DXCoil(DXCoilNum).OutletAirHumRat = (1.0 - S2PLR) * S1OutletAirHumRat + S2PLR * S12OutletAirHumRat;
```
Comment (mjwitte) [inline src/EnergyPlus/DXCoils.cc:618]: Use `PLR`, not `RuntimeFraction` to calculate leaving conditions for 2-stage operation (Coil:Cooling:DX:TwoStageWithHumidityControlMode). `RuntimeFraction` is appropriate for power consumption.

File: src/EnergyPlus/DXCoils.cc [hunk 2]
```diff
@@ -12193,14 +12192,9 @@ namespace EnergyPlus::DXCoils {
...
-                        Real64 MinAirHumRat(0.0); // set to zero because MinAirHumRat is unused argument
-                        Hfg = PsyHfgAirFnWTdb(MinAirHumRat, HSOutletAirDryBulbTemp * SpeedRatio + (1.0 - SpeedRatio) * LSOutletAirDryBulbTemp);
-                        // Average outlet HR
-                        OutletAirHumRat = InletAirHumRat - state.dataDXCoils->DXCoil(DXCoilNum).LatCoolingEnergyRate / Hfg / state.dataDXCoils->DXCoil(DXCoilNum).InletAirMassFlowRate;
-                    } else {
-                        OutletAirHumRat = (HSOutletAirHumRat * SpeedRatio) + (LSOutletAirHumRat * (1.0 - SpeedRatio));
-                    }
+                    OutletAirHumRat =
+                        ((HSOutletAirHumRat * SpeedRatio * MSHPMassFlowRateHigh) + (LSOutletAirHumRat * (1.0 - SpeedRatio) * MSHPMassFlowRateLow)) /
+                        state.dataDXCoils->DXCoil(DXCoilNum).InletAirMassFlowRate;
```
Comment (mjwitte) [inline src/EnergyPlus/DXCoils.cc]: Multispeed DX fix massflow weighting between speeds (2398ce649769b030fcaa8de84460ba0235cd64d1)

File: src/EnergyPlus/DXCoils.cc [hunk 3]
```diff
@@ -12226,6 +12220,9 @@ namespace EnergyPlus::DXCoils {
                     state.dataDXCoils->DXCoil(DXCoilNum).ElecCoolingPower = state.dataDXCoils->DXCoil(DXCoilNum).CoolingCoilRuntimeFraction * HSElecCoolingPower +
                                                          (1.0 - state.dataDXCoils->DXCoil(DXCoilNum).CoolingCoilRuntimeFraction) * LSElecCoolingPower;
                 }
+                // Now reset runtime fraction to 1.0 (because LS is running the full timestep)
+                state.dataDXCoils->DXCoil(DXCoilNum).CoolingCoilRuntimeFraction = 1.0;
```
Comment (mjwitte) [inline src/EnergyPlus/DXCoils.cc]: This inside a big `if (SpeedNum > 1 && SingleMode == 0) {` so this means it running at speed 2 or higher. (bd8aa38572816da32e4fb4f7456b24038e57dcfb)

File: src/EnergyPlus/DXCoils.cc [hunk 4]
```diff
@@ -12412,10 +12409,22 @@ namespace EnergyPlus::DXCoils {
                 }
 
                 if (FanOpMode == CycFanCycCoil) OnOffFanPartLoadFraction = PLF;
-                // outlet conditions are average of inlet and low speed weighted by CycRatio
-                OutletAirEnthalpy = LSOutletAirEnthalpy;
-                OutletAirHumRat = LSOutletAirHumRat;
-                OutletAirDryBulbTemp = LSOutletAirDryBulbTemp;
+                if (FanOpMode == ContFanCycCoil) {
+                    // outlet conditions are average of inlet and low speed weighted by CycRatio
```
Comment (mjwitte) [inline src/EnergyPlus/DXCoils.cc]: Multispeed DX fix outlet conditions at low speed (bd8aa38572816da32e4fb4f7456b24038e57dcfb)

File: src/EnergyPlus/UnitarySystem.cc [hunk 5]
```diff
@@ -2140,7 +2140,8 @@ namespace UnitarySystems {
                 this->m_IdleSpeedRatio = this->m_IdleVolumeAirRate / this->m_DesignFanVolFlowRate;
             }
 
-        } else if (this->m_CoolingCoilType_Num == DataHVACGlobals::CoilDX_MultiSpeedCooling) {
+        } else if (this->m_CoolingCoilType_Num == DataHVACGlobals::CoilDX_MultiSpeedCooling ||
+                   this->m_CoolingCoilType_Num == DataHVACGlobals::CoilDX_CoolingTwoSpeed) {
```
Comment (mjwitte) [inline src/EnergyPlus/UnitarySystem.cc]: Allow 2-speed dx to have variable speeds (ea021269050ac068e7def63e2093a434b3ed7aba)

File: src/EnergyPlus/UnitarySystem.cc [hunk 6]
```diff
@@ -2179,6 +2180,15 @@ namespace UnitarySystems {
                 this->m_IdleVolumeAirRate = this->m_MaxNoCoolHeatAirVolFlow;
                 this->m_IdleMassFlowRate = this->MaxNoCoolHeatAirMassFlow;
                 this->m_IdleSpeedRatio = this->m_IdleVolumeAirRate / this->m_DesignFanVolFlowRate;
+            } else {
+                for ( Iter = this->m_NumOfSpeedCooling; Iter > 0; --Iter ) {
+                    this->m_CoolVolumeFlowRate[ Iter ] = this->m_MaxCoolAirVolFlow * Iter / this->m_NumOfSpeedCooling;
```
Comment (mjwitte) [inline src/EnergyPlus/UnitarySystem.cc:2185]: More allow 2-speed dx to be variable speed (ea021269050ac068e7def63e2093a434b3ed7aba)

File: src/EnergyPlus/UnitarySystem.cc [hunk 7]
```diff
@@ -4414,7 +4424,13 @@ namespace UnitarySystems {
 
                         } else { // mine data from DX cooling coil
 
-                            if (thisSys.m_CoolingCoilType_Num == DataHVACGlobals::CoilDX_CoolingTwoSpeed) thisSys.m_NumOfSpeedCooling = 2;
+                            if (thisSys.m_CoolingCoilType_Num == DataHVACGlobals::CoilDX_CoolingTwoSpeed) {
```
Comment (mjwitte) [inline src/EnergyPlus/UnitarySystem.cc]: More allow 2-speed dx to be variable speed (ea021269050ac068e7def63e2093a434b3ed7aba)

**Review by mjwitte — COMMENTED** [review]
Code walkthru. Note that UnitarySystem changes overlap (duplicate?) changes in #8558.

File: src/EnergyPlus/DXCoils.cc (hunk 1 above)
Comment (rraustad) [reply src/EnergyPlus/DXCoils.cc:618]: Agreed.

File: src/EnergyPlus/DXCoils.cc (hunk 2 above)
Comment (rraustad) [reply src/EnergyPlus/DXCoils.cc]: I recall this from @lgu1234  a while back.

File: src/EnergyPlus/DXCoils.cc (hunk 3 above)
Comment (rraustad) [reply src/EnergyPlus/DXCoils.cc]: Makes sense.

File: src/EnergyPlus/DXCoils.cc (hunk 4 above)
Comment (rraustad) [reply src/EnergyPlus/DXCoils.cc]: I see in here where OutletAir* variables are calculated. Oh, there it is, in the else they are also set for cycling fan. OK then.

File: src/EnergyPlus/DXCoils.cc [hunk 8]
```diff
@@ -12436,25 +12445,7 @@ namespace EnergyPlus::DXCoils {
...
-                                                          InletAirDryBulbTemp,
-                                                          InletAirHumRat,
-                                                          OutletAirDryBulbTemp,
-                                                          OutletAirHumRat,
-                                                          state.dataDXCoils->DXCoil(DXCoilNum).SensCoolingEnergyRate,
-                                                          state.dataDXCoils->DXCoil(DXCoilNum).LatCoolingEnergyRate,
-                                                          state.dataDXCoils->DXCoil(DXCoilNum).TotalCoolingEnergyRate);
-                    }
-                }
+
```
Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc]: The above deleted code looks like it's catching up to develop.

File: src/EnergyPlus/UnitarySystem.cc (hunk 5 above)
Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc]: This is actually missing from develop? And now things line up better?

File: src/EnergyPlus/UnitarySystem.cc (hunk 6 above)
Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc:2185]: and this goes with the 2-speed coil addition to this section.

File: src/EnergyPlus/UnitarySystem.cc (hunk 5 above)
Comment (mjwitte) [reply src/EnergyPlus/UnitarySystem.cc]: @rraustad You actually made this commit a year ago. https://github.com/NREL/EnergyPlus/commit/ea021269050ac068e7def63e2093a434b3ed7aba Ring a bell?

File: src/EnergyPlus/UnitarySystem.cc [hunk 9]
```diff
@@ -7286,6 +7290,7 @@ namespace UnitarySystems {
                 }
 
                 if (state.dataUnitarySystems->unitarySys[sysNum].m_CoolingCoilType_Num == DataHVACGlobals::CoilDX_MultiSpeedCooling ||
+                    state.dataUnitarySystems->unitarySys[sysNum].m_CoolingCoilType_Num == DataHVACGlobals::CoilDX_CoolingTwoSpeed ||
                     state.dataUnitarySystems->unitarySys[sysNum].m_CoolingCoilType_Num == DataHVACGlobals::CoilDX_Cooling ||
                     state.dataUnitarySystems->unitarySys[sysNum].m_HeatingCoilType_Num == DataHVACGlobals::CoilDX_MultiSpeedHeating ||
                     state.dataUnitarySystems->unitarySys[sysNum].m_HeatingCoilType_Num == DataHVACGlobals::Coil_HeatingElectric_MultiStage ||
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: I see, you moved the report variable set up to here (but didn't get rid of the "if" in above code at 7207).

File: src/EnergyPlus/UnitarySystem.cc [hunk 10]
```diff
@@ -9077,7 +9082,10 @@ namespace UnitarySystems {
         // RETURN if the moisture load is met
         if (state.dataUnitarySystems->MoistureLoad >= 0.0 || state.dataUnitarySystems->MoistureLoad >= TempLatOutput) return;
         // Multimode does not meet the latent load, only the sensible load with or without HX active
+         // what if there is a heating load for a system using Multimode?
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: May need to create a new issue to see if this works as expected.

File: tst/EnergyPlus/unit/UnitarySystem.unit.cc [hunk 11]
```diff
@@ -840,12 +840,12 @@ TEST_F(ZoneUnitarySysTest, Test_UnitarySystemModel_factory)
...
     EXPECT_NEAR(100.0, thisSys->m_AncillaryOnPower, 0.00000001);
     EXPECT_NEAR(50.0, thisSys->m_AncillaryOffPower, 0.00000001);
-    EXPECT_NEAR(0.49388, thisSys->m_PartLoadFrac, 0.000001);
+    EXPECT_NEAR(0.488072083, thisSys->m_PartLoadFrac, 0.000001);
     Real64 totalAncillaryPower =
         thisSys->m_AncillaryOnPower * thisSys->m_PartLoadFrac + thisSys->m_AncillaryOffPower * (1.0 - thisSys->m_PartLoadFrac);
     EXPECT_NEAR(totalAncillaryPower, thisSys->m_TotalAuxElecPower, 0.00000001);
     // at PLR very near 0.5, m_TotalAuxElecPower should be very near 75 W.
-    EXPECT_NEAR(74.694, thisSys->m_TotalAuxElecPower, 0.0001);
+    EXPECT_NEAR(74.4036, thisSys->m_TotalAuxElecPower, 0.0001);
```
Comment (rraustad) [inline tst/EnergyPlus/unit/UnitarySystem.unit.cc]: small diff's in unit tests, good to go so far.

File: src/EnergyPlus/UnitarySystem.cc (hunk 5 above)
Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc]: No, but at least I'm consistent. And that whole conversation of what does idle speed anything mean is still a little confusing. I see now there is a call to DXCoils::SimDXCoilMultiSpeed. That is the correct call for 2-spd coil? I've lost track but thought is was a different function.

Comment (mjwitte) [reply src/EnergyPlus/UnitarySystem.cc]: Yes, that's fine. There are further checks on coil type in that function.
line 343 `            if (SELECT_CASE_var == CoilDX_CoolingTwoSpeed) {`

File: src/EnergyPlus/UnitarySystem.cc (hunk 9 above)
Comment (mjwitte) [reply src/EnergyPlus/UnitarySystem.cc]: Good catch - will clean that up.

**Comment by mjwitte** [comment]
@rraustad Thanks for the review.

**Comment by Myoldmopar** [comment]
@mjwitte my global branch is about to merge.  I don't anticipate major conflicts with this branch.  I believe @mitchute also has another globals branch soon to be ready to merge.  After that I'm back to trying to get the other PRs reviewed and merged for a little while, then more globals tomorrow afternoon.  All that to say, if any more conflicts arise, I'm happy to resolve them. :+1:

**Comment by mjwitte** [comment]
@Myoldmopar Much appreciated. I'm done with this for the time being. I have not looked at the diffs to make sure there's nothing ridiculous in here.

**Comment by Myoldmopar** [comment]
Pulled develop in, there was only a tiny conflict even after all those PRs.  I ran the unit test suite, and I am starting to see some SQL related unit tests fail, but they only seem to be happening when the unit tests are run serially, and it's not consistent.  I don't have time to investigate right now, so I'm just going to pretend for tonight like they aren't happening.

I'll run the regressions locally, and assuming CI is OK with my merge I'll take a peek into the diffs and we'll see if we can drop this right in tomorrow morning, then follow it up with the other coil branch.

**Comment by Myoldmopar** [comment]
I am able to reproduce the diffs locally as they are on CI.

- EIO diffs: Warmup convergence, tiny
- ERR diffs: Reduced hvac iteration warnings
- Table diffs: Small changes in peak electricity usage, generally < 1%
- ESO diffs: Values generally changed on the order of e-8, unitary system electricity off by ~1%
- MTR diffs: HVAC electricity off by less than 1%

I took the big diff math files and re-ran them with only annual results and got zero out of range math diffs.  I believe this is good to go in.  Thanks @mjwitte and @rraustad !  This is great!

**Pull request merged**